x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -794,316 +775,294 @@ private void findTokenFromRefreshToken(String refreshToken, ActionListener<Searc
     }
 
     /**
-     * Performs the actual refresh of the token with retries in case of certain exceptions that may be recoverable. The
-     * refresh involves two steps:
-     * First, we check if the token document is still valid for refresh ({@link TokenService#checkTokenDocForRefresh(Map, Authentication)}
-     * Then, in the case that the token has been refreshed within the previous 30 seconds (see
-     * {@link TokenService#checkLenientlyIfTokenAlreadyRefreshed(Map, Authentication)}), we do not create a new token document
-     * but instead retrieve the one that was created by the original refresh and return a user token and
-     * refresh token based on that ( see {@link TokenService#reIssueTokens(Map, String, ActionListener)} ).
-     * Otherwise this token document gets its refresh_token marked as refreshed, while also storing the Instant when it was
-     * refreshed along with a pointer to the new token document that holds the refresh_token that supersedes this one. The new
-     * document that contains the new access token and refresh token is created and finally the new access token and refresh token are
-     * returned to the listener.
+     * Performs the actual refresh of the token with retries in case of certain exceptions that may be recoverable. The refresh involves two
+     * steps: First, we check if the token document is still valid for refresh. Then, in the case that the token has been refreshed within
+     * the previous 30 seconds, we do not create a new token document but instead retrieve the one that was created by the original refresh
+     * and return an access token and refresh token based on that. Otherwise this token document gets its refresh_token marked as refreshed,
+     * while also storing the Instant when it was refreshed along with a pointer to the new token document that holds the refresh_token that
+     * supersedes this one. The new document that contains the new access token and refresh token is created and finally the new access
+     * token and refresh token are returned to the listener.
      */
     private void innerRefresh(String tokenDocId, Map<String, Object> source, long seqNo, long primaryTerm, Authentication clientAuth,
```

Comment:
I think we need to ping @kobelb here and maybe help kibana get a list of potential messages for the `InvalidGrantException`s we throw (if these have changed) as this affects the way that Kibana treats auth errors wrt to refreshing tokens. 

In a follow up, we can probably sit down and define a list of short error codes that we can use instead of descriptive messages that can change in the future

Reply:
Please! At this moment, we're looking at the `error_description` of the HTTP response body for the following two strings to determine if the error represents an invalid refresh token:

- token has already been refreshed
- refresh token is expired

/cc @azasypkin

Reply:
I think it is safe to assume HTTP error code 400 is an invalid refresh token. If you find it otherwise, I would suggest we treat it as a bug. Would this be satisfiable for you @kobelb ?

There are many more messages to account for (which were not introduced/changed by this PR) if you insist on parsing the `error_description` in the Exception object. Not even all failure paths will return an `ElasticsearchException` - that has the `error_description` field.

Reply:
> I think it is safe to assume HTTP error code 400 is an invalid refresh token. If you find it otherwise, I would suggest we treat it as a bug.

Currently Kibana automatically re-initiates SAML handshake only if refreshing fails with one of the errors mentioned above, otherwise we just return error. @albertzaharovits do you think we should re-initiate SAML handshake in Kibana as long as we get `400` irrespective to the underlying reason? If brand new SAML session is the only recovery mechanism for all `400` errors we get while trying to refresh then it seems there is no need to inspect `error_description` on our side.

Reply:
@azasypkin I think so yes. We should strive to make 400 errors client errors, so they should not be due to ES instability, to the effect that a retry with the same contents will never be accepted. If Kibana decides to re-initiate the SAML flow, or present the error to the user and abort, is a different story IMO. That being said, we have not taken great care that 400 errors are honored as described, I will take on me the responsibility to look over the code again and see that 400 and 500s are consistent.

Reply:
The one place where this is not happening is 
https://github.com/albertzaharovits/elasticsearch/blob/0d4872665adefb8c4db2d4fedcd36dc446131d02/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java#L383

where we return a 500 when we can't find the token document based on the Bearer token that was provided. This made sense when the Bearer token was the encrypted serialized version of the UserToken, but not anymore - I have this tracking in https://github.com/elastic/elasticsearch/issues/38866 and plan to address this soon, now that this PR is merged. 


I totally agree with Albert here, if you get a `400` while refreshing then it means that the the refresh token can't be used, no matter what that reason is. I think we should re-inititate the authentication flow from Kibana when this happens, as I can't think of a case where the failure to refresh should cause a user to see an error in Kibana.

Reply:
Okay, sounds good, filed https://github.com/elastic/kibana/issues/33646